### PR TITLE
`arrow::compute::kernels::temporal` should support nanoseconds 

### LINKED ISSUE #2995: arrow::compute::kernels::temporal should support nanoseconds
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`arrow::compute::kernels::temporal` should support nanoseconds in similar way it supports seconds

**Describe the solution you'd like**

Add functions to extracts the nanoseconds of a given temporal array as an array of integers

**Describe alternatives you've considered**

None
**Additional context**

None

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #2995.

# Rationale for this change
`arrow::compute::kernels::temporal` should support nanoseconds to let calculate nanos, millis, micros

# What changes are included in this PR?

Support nano seconds for temporal
# Are there any user-facing changes?

Added public function `nanosecond`

### CODE REVIEW

**Comment by comphead** [comment]
@waitingkuo @tustvold please review the PR whenever you have time.

File: arrow/src/compute/kernels/temporal.rs [hunk 1]
```diff
@@ -717,49 +717,65 @@ where
     T: ArrowTemporalType + ArrowNumericType,
     i64: From<T::Native>,
 {
-    second_generic::<T, _>(array)
+    second_fraction_generic::<T, _, _>(array, "second", |t| t.second() as i32)
 }
 
-/// Extracts the seconds of a given temporal array as an array of integers
-pub fn second_generic<T, A: ArrayAccessor<Item = T::Native>>(
```
Comment (tustvold) [inline arrow/src/compute/kernels/temporal.rs:724]: I think we have lost this function?

File: arrow/src/compute/kernels/temporal.rs [hunk 2]
```diff
@@ -717,49 +717,65 @@ where
...
+pub fn nanosecond<T>(array: &PrimitiveArray<T>) -> Result<Int32Array>
+where
+    T: ArrowTemporalType + ArrowNumericType,
+    i64: From<T::Native>,
+{
+    second_fraction_generic::<T, _, _>(array, "nanosecond", |t| t.nanosecond() as i32)
+}
+
+/// Extracts the seconds fraction of a given temporal array as an array of integers
+pub fn second_fraction_generic<T, A: ArrayAccessor<Item = T::Native>, F>(
```
Comment (tustvold) [inline arrow/src/compute/kernels/temporal.rs:733]: Perhaps we could call this something like unary_timestamp or something? I would also not make it public, so we can alter it down the line

File: arrow/src/compute/kernels/temporal.rs [hunk 3]
```diff
@@ -717,49 +717,65 @@ where
...
         }
-        dt => second_internal::<T, A>(array, &dt),
+        dt => second_fraction_internal::<T, A, _>(array, &dt, name, op),
     }
 }
 
-/// Extracts the seconds of a given temporal array as an array of integers
-fn second_internal<T, A: ArrayAccessor<Item = T::Native>>(
+/// Extracts the seconds fraction of a given temporal array as an array of integers
+fn second_fraction_internal<T, A: ArrayAccessor<Item = T::Native>, F>(
```
Comment (tustvold) [inline arrow/src/compute/kernels/temporal.rs:752]: It occurs to me that this function can probably be implemented more efficiently using try_unary, probably not something for this PR though

**Review by tustvold — COMMENTED** [review]
Left some comments, I think ideally we would avoid making a breaking API change as part of this

**Comment by comphead** [comment]
Thanks @tustvold for the feedback.

There is no breaking change, I have made `second_generic` and `second_internal` functions more generic so they can serve for both nanosecond and second. `Second_generic` and `second_internal` have been renamed to `second_fraction_generic` and `second_fraction_internal` respectively. I agree about better naming, appreciate if you can provide the name :)

File: arrow/src/compute/kernels/temporal.rs (hunk 2 above)
Comment (waitingkuo) [reply arrow/src/compute/kernels/temporal.rs:733]: i suggest that we could either
1. separate them (aka second_generic, and nano_generic), keep it consistent wither other operator
2. make it more general, combines not only second, nanosecond. (e.g. `second_fraction_generic::<T, _, _>(array, "minute", |t| t.minute() as i32` should work for `minute`)

**Comment by comphead** [comment]
@waitingkuo thanks for your feedback. Renamed functions to `time_fraction_generic`, now it also works for minutes. Ideally it can work for hour, but hour by some reason has more conditions than minute, second, nanosecond. I can check hour in separate PR

File: arrow/src/compute/kernels/temporal.rs [hunk 4]
```diff
@@ -717,49 +674,65 @@ where
...
+pub fn nanosecond<T>(array: &PrimitiveArray<T>) -> Result<Int32Array>
+where
+    T: ArrowTemporalType + ArrowNumericType,
+    i64: From<T::Native>,
+{
+    time_fraction_generic::<T, _, _>(array, "nanosecond", |t| t.nanosecond() as i32)
+}
+
+/// Extracts the time fraction of a given temporal array as an array of integers
+pub fn time_fraction_generic<T, A: ArrayAccessor<Item = T::Native>, F>(
```
Comment (tustvold) [inline arrow/src/compute/kernels/temporal.rs:690]: Can we make this private

**Review by tustvold — COMMENTED** [review]
This PR seems to now remove the public second_generic and minute_generic functions. This makes it a breaking API change as downstreams using these methods will no longer work without modification. Perhaps we could rollback this change?

I would like to remove these _generic kernels, and replace them with dyn variants like we have for other kernels, but I think we should do this as a separate PR for all of them - I will create a ticket flr this

Edit: created https://github.com/apache/arrow-rs/issues/3004 perhaps you might like to work on this?

File: arrow/src/compute/kernels/temporal.rs (hunk 4 above)
Comment (comphead) [reply arrow/src/compute/kernels/temporal.rs]: Made private. Returned public `second_generic` and `minute_generic` methods as well as tests for them.

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Thank you

**Comment by waitingkuo** [comment]
LGTM thank you @comphead  @tustvold

**Pull request merged**